crypto/x509/x509_vfy.c

```diff
@@ -854,13 +859,12 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
          * we'll accept X509_TRUST_UNTRUSTED when not self-signed.
          */
         trust = X509_check_trust(mx, ctx->param->trust, 0);
-        if (trust == X509_TRUST_REJECTED) {
-            X509_free(mx);
+        if (trust == X509_TRUST_REJECTED)
             goto rejected;
-        }
 
         /* Replace leaf with trusted match */
-        (void) sk_X509_set(ctx->chain, 0, mx);
+        if (!sk_X509_set(ctx->chain, 0, mx) || !X509_up_ref(mx))
```

Comment:
This up_ref may be too late, you've already freed the returned list (stack) from which this was chosen. Yes, in practice there's likely some other structure that's still holding on to the certificate, but that guarantee may be weak.

Reply:
Right, `X509_up_ref()` needs to be called before calling `sk_X509_set()`,
and the result of the latter can be ignored (as this was done before this PR).
Corrected.

Reply:
Sorry, you have multiple overlapping PRs in flight, and it is not easy to keep track of which ones do what.  This is not the PR where you run into trouble with the up_ref.  Rather it is the one where you fail to up_ref the cert in lookup_cert_match(), where it is not optional to do so.  And with that done, you don't need to do it here.